Update/restructure live data (#907)
MapsPeople/docs: src/content/map/live-data/live-data-intro.md

</mi-tab-panel>
<mi-tab-panel id="ios">

<!-- Known Issues -->
{% include "src/content/shared/known-issues-ios.md" %}

## Enable Live Data in Your App in iOS

To enable Live Data in an application, a subscription to one or more Topics is needed. Once subscribed, the application can be notified about changes and decide what to do. The application is in control of what should happen upon receiving live data updates, and the MapsIndoors SDKs provide mechanisms to efficiently make updates to the map representation of locations.

The central class to carry out the subscription tasks is the `LiveDataManager`, but the `MPMapControl` class also has some convenience methods that abstract the subscription layer, trading off some granular control.

### Enable Live Data through MPMapControl in iOS

Enabling of Live Data through `MPMapControl` can be done as simple as calling `MPMapControl.enableLiveData()` with a Domain Type.

```swift
  self.mapControl = MPMapControl.init(map: self.map!)

  self.mapControl.enableLiveData(MPLiveDomainType.position)
```

> In the example we are enabling Live Data for the Domain Type "Occupancy". Internal processes will determine which topics are relevant for subscription based on where the map is situated. A default rendering mechanism will also alter the appearance of the relevant locations on the map. As a consequence, the SDK will set [custom display rules]({{ site.url }}/content/map/map-styling/map-styling-ios/) for this rendering. Adding your own or resetting display rules while Live Data is enabled with default rendering may break the rendering for the current `MPMapControl` instance. Hence, you should not use custom display rules unless you are [handling the rendering of Live Data](#rendering-live-data-locations) by your own.

Reviewer: ```suggestion
In the example we are enabling Live Data for the Domain Type "Occupancy". Internal processes will determine which topics are relevant for subscription based on where the map is situated. A default rendering mechanism will also alter the appearance of the relevant locations on the map. As a consequence, the SDK will set [custom display rules]({{ site.url }}/content/map/map-styling/map-styling-ios/) for this rendering. Adding your own or resetting display rules while Live Data is enabled with default rendering may break the rendering for the current `MPMapControl` instance. Hence, you should not use custom display rules unless you are [handling the rendering of Live Data](#rendering-live-data-locations) on your own.
```

> Note that using the `enableLiveData()` methods on `MPMapControl` has some limitations and is thereby not suitable for all use cases.

Reviewer: ```suggestion
Note that using the `enableLiveData()` methods on `MPMapControl` has some limitations and is therefore not suitable for all use cases.
```

- Since `MPMapControl` will try to determine relevant Live Data subscriptions based on where the map is currently situated, it might not detect Live Data updates of the Position domain representing moving objects entering the visible region of the map.

> - Since `MPMapControl` does not know which Live Updates that are relevant to show it will need to subscribe to all Live Data in the visible region, which, depending on your amount of Live Data, may or may not lead to app performance implications.

Reviewer: ```suggestion
- Since `MPMapControl` does not know which Live Updates are relevant to show, it will need to subscribe to all Live Data in the visible region, which, depending on your amount of Live Data, may or may not impact app performance.
```

2. On a map-specific level through `MPMapControlDelegate` or `MPMappedLocationUpdateHandler`, which is suitable in scenarios where the map is the context for the user's actions, for example when browsing the map for an available meeting room nearby.

To get Live Updates on a general level the `MPLiveDataManagerDelegate` protocol method `didReceive(_ liveUpdate: MPLiveUpdate)` must be implemented:

```swift
extension MyClass : MPLiveDataManagerDelegate {
    func didReceive(_ liveUpdate: MPLiveUpdate) {
        let liveUpdate = location.getLiveUpdate(forDomainType: MPLiveDomainType.occupancy)
        ...
    }
}
```

To get Live Updates on a map-specific level the `MPMapControlDelegate` protocol method `willUpdateLocationsOnMap(locations: [MPLocation])` must be implemented:

```swift
extension MyClass : MPMapControlDelegate {
    func willUpdateLocationsOnMap(locations: [MPLocation]) {
        for location in locations {
            let liveUpdate = location.getLiveUpdate(forDomainType: MPLiveDomainType.occupancy)
            ...
        }
    }
}
```

### Handling State Changes and Errors in iOS

> In order to get notified about state changes and errors happening in the Live Data Manager, `MPLiveDataManagerDelegate` utilizes other methods as shown in the example below:

Reviewer: ```suggestion
In order to get notified about state changes and errors happening in the Live Data Manager, `MPLiveDataManagerDelegate` provides methods as shown in the example below:
```

```swift
extension MyClass : MPLiveDataManagerDelegate {

    func didUpdate(_ state: MPLiveDataManagerState) {
        print("Manager state changed to \(state)")
    }

    func didSubscribe(_ topic: MPLiveTopicCriteria) {
        print("Started Live Updates for \(topic.topicString)")
    }

    func didUnsubscribe(_ topic: MPLiveTopicCriteria) {
        print("Stopped Live Updates for \(topic.topicString)")
    }

    func onSubscriptionError(_ error: Error, topic: MPLiveTopicCriteria) {
        print("Could not subscribe Live Updates for \(topic.topicString)")
    }

    func onError(_ error: Error) {
        print("We got an error")
    }

}
```

> Live Updates are of course dependent on network connectivity, so the Live Data Manager will try to recover from common errors like network dropout. On the contrary, the Live Data Manager will not try to recover from subscription errors alone as this could be caused by a non-existing topic for a given Dataset, thus it does not make sense retrying the failing subscription.

Reviewer: ```suggestion
Live Updates are of course dependent on network connectivity, so the Live Data Manager will try to recover from common errors like network dropout. However, the Live Data Manager will not try to recover from subscription errors alone as this could be caused by a non-existing topic for a given Dataset, thus it does not make sense retrying the failing subscription.
```
